server/src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java

```diff
@@ -131,7 +128,7 @@ public synchronized FetchResult<T> fetchData(DiscoveryNodes nodes, Set<String> i
         }
 
```

Comment:
nit: ideally we'd also assert consistency between `fetchingCount` and `cache` here (or inside the brace one line up).

```suggestion
assert assertFetchingCountConsistent();
```

Reply:
++ thanks, yes that'd make sense. I'll do that in a follow up (and apply it in `main` too)

Reply:
see https://github.com/elastic/elasticsearch/pull/96553